README.md
# Introduction
The Open Source graphical user interface (GUI) for a SpO2 simulator, named as UAO_SpO2_Sim, it's a project that was born in 2017 in the Research Group in Biomedical Engineering, GBIO, at the Autonomous University of the West (Universidad Autónoma de Occidente), in Cali Colombia.

An easy-quick review of this project could be found in the IEEExplore by the name "Development of a Low-Cost Pulse Oximeter Simulator for Educational Purposes", link:

* https://ieeexplore.ieee.org/document/8564698
* DOI: 10.1109/ANDESCON.2018.8564698
* Reference: K. Machado-Gamboa and A. Gonzalez-Vargas, "Development of a Low-Cost Pulse Oximeter Simulator for Educational Purposes," 2018 IEEE ANDESCON, Santiago de Cali, Colombia, 2018, pp. 1-6. doi: 10.1109/ANDESCON.2018.8564698

This project contains software and hardware. You will find in this Github just the coding related to the software developed or the GUI. A quick view of the software results will be available in an article that will be soon publish in the IEEExplore by the name "Development of a low-cost pulse oximeter simulator for educational purposes". The results in detail, containing also the hardware, will be available for the public in general in the online repository of Universidad Autonoma de Occidente, Cali, Colombia, by the name "DESARROLLO E IMPLEMENTACIÓN DE UN SIMULADOR DE PULSIOXIMETRÍA PARA USO ACADÉMICO". The link of the repository is:

* http://red.uao.edu.co/

## Software Requirements

The quickest way is to install the last version of Anaconda from its official site
* https://docs.anaconda.com/anaconda/install/

This project was developed using Spyder 3.2.4 and Qt Designer 5.2.1 So any higher version should work properly.
Secondly, we need to ensure that we have installed the next python libraries:

* pyqtgraph   ---->    (Not included in Anaconda)
* PyQt5
* multiprocessing
* scipy
* collections
* numpy
* sys
* time

Please make sure all libraries get install in Python 3. Note that all the necessary libraries are included in Anaconda except pyqtgraph. To install pyqtgraph
in Anaconda environment you can try typing: `conda install -c anaconda pyqtgraph`; or visit http://www.pyqtgraph.org/

## Files Description
* "mainWindowPPG.py" contains the main code of the GUI
* "mainWindowPPG.ui" contains the GUI developed in the software Qt. 
* "UAOspo2sim.py" contains the main code but to be run with the Rpi 3 B+ and the ADC PCF8591
* "curvesHB.mat" contains a matrix in which each row is a vector that displays a curve. This is use for some graphical effects. 

## Summary of the main idea of the GUI 
Understanding the Pulse Oximetry technique requires the a set of knowledge, including:
- light absorption phenomenon of two beams with wavelength between 640 nm and 950 nm in hemoglobin Hb & oxyhemoglobin HbO2.
- The generation of red and infrared photoplethysmographic signals components after passing through the skin bed as a finger.
- The comprehension of ratio and AC and DC components of these red and infrared photoplethysmographic signals.
- The calculation of SpO2 (%) in function of ratio

With this GUI we intend to visually include these set of knowledge within the interface. This fact makes the operator of the simulation to visualize this figures regarding the pulse oximeter techniques anytime he/she operates the simulator.

## GUI preview
Once everything gets installed you will display a GUI as the picture bellow:
![interfaz](https://user-images.githubusercontent.com/15948497/47217427-37407780-d3a0-11e8-9995-735536e6769b.jpg)


changelog.md
# ChangeLog File
This file contains a chronologically ordered list of notable changes for each version of a project since the 16 of January 2021.

[0.2.0] - 2021-01-16
Spo2 signal flipped and minor code refactors

[0.1.0] - 2017-06-20
First version of the simulator.


UAOspo2sim.py
"""
UAO SpO2 Simulator GUI - Hardware Required !
-------------------------------------
Created on Thu Dec 21 12:07:49 2017
by the author: Kevin Machado Gamboa
Contact: ing.kevin@hotmail.com
Modified on Sun Oct 11 10:29:39 2020
------------------------------------
REFERENCES: 
    [1] Development of a Low-Cost Pulse Oximeter Simulator for Educational Purposes
    https://github.com/kevinmgamboa/UAO_SpO2_Sim
    [2] Sebastian Sepulveda - plot data from a function in real time 
    https://github.com/ssepulveda/RTGraph/tree/oldRTGraph
"""
import sys
import numpy as np
from time import time
import scipy.io as sio
import PCF8591v2 as ADC
from collections import deque
from multiprocessing import Queue
# Librery for the management of Qt v5. UI platform
from PyQt5.uic import loadUi
from PyQt5 import QtCore
from PyQt5.QtWidgets import QApplication, QMainWindow, QSlider

ADC.setup(0x48, 0x4c)
# @brief Buffer size for the data (number of points in the plot)
N_SAMPLES = 100
# @brief Update time of the plot, in ms
PLOT_UPDATE_TIME = 10
# @brief Point to update in each redraw
PLOT_UPDATE_POINTS = -1

class mainWindow(QMainWindow):
    def __init__(self):
        #Inicia el objeto QMainWindow
        QMainWindow.__init__(self)
        # Loads an .ui file & configure UI
        loadUi("mainWindowPPG.ui",self)
        self.setupUI()   
        # Shared variables, initial values
        self.queue = Queue(N_SAMPLES)
        self.dataR = deque([], maxlen=N_SAMPLES)
        self.dataIR = deque([], maxlen=N_SAMPLES)
        self.TIME = deque([], maxlen=N_SAMPLES)
        
        self._plt = None
        self._plt_2 = None
        self._Rplt = None
        self._Rplt_2 = None
        
        self._timer_plot = None
        self.plot_colors = ['#0072bd', '#d95319', '#bd0000']
        # Spo2 signal initial parameters
        self.timestamp = 0.0
        self.ampR = 0.4      # amplitud for Red signal
        self.ampIR = 0.270      # amplitud for InfraRed signal
        self.minR = 1.45   # Desplacement from zero for Red signal
        self.minIR = 1.45   # Desplacement from zero for Red signal

        self._configure_plot()
        # UI connectors
        self.HRsl.valueChanged.connect(self.spo2sl_change)
        self.spo2sl.valueChanged.connect(self.spo2sl_change)        
        # Configurations
        
        self._enable_ui(True)        
        self._configure_timers()
        self.buttons()
        self._initial(self.ampR, self.ampIR)
        self.spo2sl_change()

    def start(self):
        """
        This function works when the start button is clicked
        It generates a t0 time and activates the Qt timer which connects to update_plot
        :return:
        """
        self.stop()
        self._plt_2.clear()
        self._enable_ui(False)
        self.timestamp= time()
        self._timer_plot.start(PLOT_UPDATE_TIME)
        
    def stop(self):
        """
        This function works when the stop button is clicked
        it stop the timer and resets the buffers
        """
        self._initial(self.ampR, self.ampIR)
        ADC.write(0,0)
        self._enable_ui(True)
        self._timer_plot.stop()  
        self.reset_buffers()
    
    def _update_plot(self):
        """
        Updates and redraws the graphics in the plot.
        :return:
        """
        # Geting heart rate
        HR = float(self.HRsl.value())
        # generates the time
        self.tPPG = time() - self.timestamp
        self.sR, self.sIR = self.ppg_parameters(self.minR, self.ampR, self.minIR, self.ampIR, self.tPPG, HR)
        yR = 1 * ( ( (self.sR - 0)/(3.0 - 0) ) * 255)
        yIR = 1 * ( ( (self.sIR - 0)/(3.0 - 0) ) * 255)

        # store data into variables 
        self.TIME.append(self.tPPG)
        self.dataR.append(self.sR)
        self.dataIR.append(self.sIR)

        ADC.write(yR,yIR)
        # Draw new data
        self._plt_2.clear()
        self._plt_2.plot(x=list(self.TIME)[-PLOT_UPDATE_POINTS:], y=list(self.dataR)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[1])
        self._plt_2.plot(x=list(self.TIME)[-PLOT_UPDATE_POINTS:], y=list(self.dataIR)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[0])
                
    def setupUI(self):
        """
       Configures everything regarding the UI
        """
        # Defult Heart Rate configuration
        self.HRsl.setMaximum(250)
        self.HRsl.setMinimum(50)
        self.HRsl.setValue(80)
        self.HRsl.setTickPosition(QSlider.TicksBelow)
        self.HRsl.setTickInterval(1)
        # Defult SpO2 configuration
        self.spo2sl.setMaximum(100)
        self.spo2sl.setMinimum(50)
        self.spo2sl.setValue(100)
        self.spo2sl.setTickPosition(QSlider.TicksBelow)
        self.spo2sl.setTickInterval(1)
    
    def _configure_plot(self):
        """
        Configures specific elements of the PyQtGraph plots.
        :return:
        """
        self.plt.setBackground(background=None)
        self.plt.setAntialiasing(True)
        self._plt = self.plt.addPlot(row=1, col=1)
        self._plt.setLabel('bottom', "Time", "s")
        self._plt.setLabel('left', "Amplitude", "Volt")
        self._plt.showGrid(x=False, y=True)
        
        self.plt_2.setBackground(background=None)
        self.plt_2.setAntialiasing(True)
        self._plt_2 = self.plt_2.addPlot(row=1, col=1)
        self._plt_2.setLabel('bottom', "Time", "s")
        self._plt_2.setLabel('left', "Amplitude", "Volt")
        self._plt_2.showGrid(x=False, y=False)
        
        self.Rplt.setBackground(background=None)
        self.Rplt.setAntialiasing(True)
        self._Rplt = self.Rplt.addPlot(row=1, col=1)
        self._Rplt.setLabel('bottom', "R-value")
        self._Rplt.setLabel('left', "%SpO2")
        self._Rplt.showGrid(x=False, y=True)
        
        self.Rplt_2.setBackground(background=None)
        self.Rplt_2.setAntialiasing(True)
        self._Rplt_2 = self.Rplt_2.addPlot(row=1, col=1)
        self._Rplt_2.setLabel('bottom', "Wavelength", "*nm")
        self._Rplt_2.setLabel('left', "Absorbance")
        self._Rplt_2.showGrid(x=False, y=True)
    
    def _configure_timers(self):
        """
        Configures specific elements of the QTimers.
        :return:
        """
        self._timer_plot = QtCore.QTimer(self) # gives _timer_plot the attribute of QtCore.QTimer
        self._timer_plot.timeout.connect(self._update_plot)  # connects with _update_plot method  
    
    def ppg_parameters(self, minR, ampR, minIR, ampIR, t, HR):
        """
        Store the function of two signals - e.g PPG Red and Infrared channel signals
        We can also put here a sine, cosine, etc.
        """
        f= HR * (1/60)
        # Spo2 Red signal function
        self.sR= minR + ampR * (0.5*np.sin(2*np.pi*t*f) + 0.22*np.sin(2*np.pi*t*2*f+40))
        # Spo2 InfraRed signal function
        self.sIR= minIR + ampIR * (0.5*np.sin(2*np.pi*t*f) + 0.22*np.sin(2*np.pi*t*2*f+40))
        
        return self.sR, self.sIR
    
    def _initial(self, ampR, ampIR):
        """
        contain the initial figure in the UI
        """
        HR = self.HRsl.value()
        self.showHR.setText(str(HR))
        
        t=np.linspace(0.2,0.8,100);
        sR,sIR = self.ppg_parameters(self.minR, ampR, self.minIR, ampIR, t, 200)
        
        curves=sio.loadmat('LinearVectors');
        
        x_new=curves['x_new']
        y_new=curves['y_new']
        
        x_HBo2=x_new[0,:]
        y_HBo2=y_new[0,:]
        
        x_HB=x_new[100,:]
        y_HB=y_new[100,:]

        spo2value = self.spo2sl.value()

        self._plt.clear()
        self._plt.plot(x=list(t)[-PLOT_UPDATE_POINTS:], y=list(sR)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[1])
        self._plt.plot(x=list(t)[-PLOT_UPDATE_POINTS:], y=list(sIR)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[0])
        
        self._Rplt_2.clear()
        self._Rplt_2.plot(x=list(x_HBo2), y=list(y_HBo2), pen=self.plot_colors[0])
        self._Rplt_2.plot(x=list(x_HB), y=list(y_HB), pen=self.plot_colors[2])
        self._Rplt_2.plot(x=list(x_new[(100-(spo2value)),:]), y=list(y_new[(100-(spo2value)),:]), pen=self.plot_colors[1])
    
    def buttons(self):
        """
        Configures the connections between signals and UI elements.
        """
        self.startButton.clicked.connect(self.start)
        self.stopButton.clicked.connect(self.stop)
    
    def reset_buffers(self):
        """
        Clear everything into the vectors that have the signals
        """
        self.dataR.clear()
        self.dataIR.clear()
        self.TIME.clear()
        
    def spo2sl_change(self):
        """
        Change the value of the SpO2 when movind the slider.
        It also have the list of SpO2 values vs the R value
        """
        spo2value = self.spo2sl.value()
        self.showSpo2.setText(str(spo2value))
        # Change the numbers of the equation
        spO2 = [100,99,98,97,96,95,94,93,92,91,90,89,88,87,86,85,84,83,82,81,80,79,78,77,76,75,74,73,72,71,70,69,68,67,66,65,64,63,62,61,60,59,58,57,56,55,54,53,52,51,50]
        R = [0.50,0.55,0.60,0.64,0.66,0.70,0.71,0.72,0.73,0.75,0.76,0.77,0.78,0.80,0.81,0.82,0.83,0.84,0.85,0.86,0.87,0.88,0.89,0.90,0.91,0.92,0.93,0.94,0.95,0.96,0.97,0.98,0.99,1.00,1.01,1.00,1.05,1.11,1.12,1.16,1.19,1.25,1.27,1.32,1.33,1.35,1.39,1.43,1.47,1.52,1.50]
        Ri = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
        Ri[spO2.index(spo2value)] = R[spO2.index(spo2value)]

        # R-IR values & SpO2
        rR = [0.3,0.4,0.4,0.4,0.4,0.3,0.3,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.5,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4,0.4]        
        IR = [1.0,0.9,0.7,1.0,0.9,0.7,0.7,0.6,0.6,0.58,0.57,0.54,0.52,0.5,0.5,0.48,0.47,0.46,0.45,0.445,0.44,0.43,0.42,0.39,0.4,0.39,0.38,0.38,0.37,0.35,0.36,0.35,0.35,0.34,0.34,0.34,0.33,0.32,0.32,0.31,0.3,0.3,0.3,0.29,0.29,0.28,0.28,0.27,0.27,0.26,0.25]

        self.ampR = rR[spO2.index(spo2value)]
        self.ampIR = IR[spO2.index(spo2value)]
        
        self.showR.setText(str(round(self.ampR/self.ampIR,2)))
        self.showRac.setText(str(round(self.ampR,2)))
        self.showIRac.setText(str(round(self.ampIR,2)))
        self.showDC.setText(str(round(self.minR,2)))

        self._initial(self.ampR, self.ampIR)

        self._Rplt.clear()
        self._Rplt.plot(x=list(R), y=list(spO2), pen=self.plot_colors[0])
        self._Rplt.plot(x=list(Ri), y=list(spO2), pen=self.plot_colors[0],  symbolBrush=(255,0,0))
        
    def _enable_ui(self, enabled):
        """
        Enable touching the buttons in the UI
        """
        self.startButton.setEnabled(enabled)
        self.spo2sl.setEnabled(enabled)
        self.HRsl.setEnabled(enabled)
        self.showSpo2.setEnabled(enabled)
        self.stopButton.setEnabled(not enabled)
        
#Instancia para iniciar una aplicacion en windows
app = QApplication(sys.argv)
#debemos crear un objeto para la clase creada arriba
_mainWindow = mainWindow()
    #muestra la ventana
_mainWindow.show()
    #ejecutar la aplicacion
app.exec_()


mainWindowPPG.py
"""
UAO SpO2 Simulator GUI - No Hardware Required !
-------------------------------------
Created on Thu Dec 21 12:07:49 2017
by the author: Kevin Machado Gamboa
Contact: ing.kevin@hotmail.com
Modified on Sun Oct 11 10:29:39 2020
------------------------------------
REFERENCES: 
    [1] Development of a Low-Cost Pulse Oximeter Simulator for Educational Purposes
    https://github.com/kevinmgamboa/UAO_SpO2_Sim
    [2] Sebastian Sepulveda - plot data from a function in real time 
    https://github.com/ssepulveda/RTGraph/tree/oldRTGraph
    [3] Qt for Python - https://doc.qt.io/qtforpython/
"""
# -----------------------------------------------------------------------------
#                             Libraries Needed
# -----------------------------------------------------------------------------
import sys
from typing import Any, Union, List

import numpy as np
from time import time
import scipy.io as sio
# import PCF8591v2 as ADC
from collections import deque
from multiprocessing import Queue
# Librery for the management of Qt v5. UI platform
from PyQt5.uic import loadUi
from PyQt5 import QtCore
from PyQt5.QtWidgets import QApplication, QMainWindow, QSlider

# -----------------------------------------------------------------------------
#                              GUI Development
# -----------------------------------------------------------------------------
# @brief Buffer size for the data (number of points in the plot)
N_SAMPLES = 100
# @brief Update time of the plot, in ms
PLOT_UPDATE_TIME = 1
# @brief Point to update in each redraw
PLOT_UPDATE_POINTS = -1


class mainWindow(QMainWindow):

    def __init__(self):
        # Initiate QMainWindow Object
        QMainWindow.__init__(self)
        # Loads an .ui file & configure UI
        loadUi("mainWindowPPG.ui", self)
        self.setupUI()
        # ---------------------------------------------------------- # 
        #               Shared variables, initial values
        # ---------------------------------------------------------- #
        self.dataR = deque([], maxlen=N_SAMPLES)
        self.dataIR = deque([], maxlen=N_SAMPLES)
        self.TIME = deque([], maxlen=N_SAMPLES)

        self._timer_plot = None
        self.plot_colors = ['#0072bd', '#d95319', '#bd0000']
        # Spo2 Red & InfraRed signal
        #self.sR, self.sIR = list[float], list[float]
        # Spo2 signal initial parameters
        self.timestamp = 0.0
        self.ampR = 0.4  # amplitude for Red signal
        self.ampIR = 0.270  # amplitude for InfraRed signal
        self.minR = 1.45  # Displacement from zero for Red signal
        self.minIR = 1.45  # Displacement from zero for Red signal
        # ---------------------------------------------------------- # 
        #                Running GUI Configurations
        # ---------------------------------------------------------- #
        self._configure_plot()
        # Sliders
        self.HRsl.valueChanged.connect(self.spo2sl_change)
        self.spo2sl.valueChanged.connect(self.spo2sl_change)
        # Other Configurations        
        self._enable_ui(True)
        self._configure_timers()
        self.buttons()
        self._initial(self.ampR, self.ampIR)
        self.spo2sl_change()

    ## -----------------------------------
    ##    SpO2 Parameters & Plotting
    ## -----------------------------------
    def ppg_parameters(self, minR, ampR, minIR, ampIR, t, HR):
        """
        Store the function of two signals - e.g PPG Red and Infrared channel signals
        We can also put here a sine, cosine, etc.
        """
        f = HR * (1 / 60)
        # Spo2 Red signal function
        self.sR = minR + ampR * (0.05 * np.sin(2 * np.pi * t * 3 * f) + 0.4 * np.sin(2 * np.pi * t * f) +
                                 0.22 * np.sin(2 * np.pi * t * 2 * f + 45))
        # self.sR = minR + ampR * (0.5 * np.sin(2 * np.pi * t * f) + 0.22 * np.sin(2 * np.pi * t * 2 * f + 40))
        # Spo2 InfraRed signal function
        self.sIR = minIR + ampIR * (0.05 * np.sin(2 * np.pi * t * 3 * f) + 0.4 * np.sin(2 * np.pi * t * f) +
                                    0.22 * np.sin(2 * np.pi * t * 2 * f + 45))
        # self.sIR = minIR + ampIR * (0.5 * np.sin(2 * np.pi * t * f) + 0.22 * np.sin(2 * np.pi * t * 2 * f + 40))

        return self.sR, self.sIR

    def spo2sl_change(self):
        """
        Change the value of the SpO2 when moving the slider.
        It also have the list of SpO2 values vs the R value
        """
        spo2value = self.spo2sl.value()
        self.showSpo2.setText(str(spo2value))
        # sp02 list from 100 to 50
        sp02 = list(range(50, 101))[::-1]

        R = [0.50, 0.55, 0.60, 0.64, 0.66, 0.70, 0.71, 0.72, 0.73, 0.75, 0.76, 0.77, 0.78, 0.80, 0.81, 0.82, 0.83,
             0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.90, 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.00,
             1.01, 1.00, 1.05, 1.11, 1.12, 1.16, 1.19, 1.25, 1.27, 1.32, 1.33, 1.35, 1.39, 1.43, 1.47, 1.52, 1.50]

        Ri = [0] * 51

        Ri[sp02.index(spo2value)] = R[sp02.index(spo2value)]

        # R-IR values & SpO2
        rR = [0.3, 0.4, 0.4, 0.4, 0.4, 0.3, 0.3, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.5, 0.4, 0.4,
              0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4,
              0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4, 0.4]

        IR = [1.0, 0.9, 0.7, 1.0, 0.9, 0.7, 0.7, 0.6, 0.6, 0.58, 0.57, 0.54, 0.52, 0.5, 0.5, 0.48, 0.47, 0.46, 0.45,
              0.445, 0.44, 0.43, 0.42, 0.39, 0.4, 0.39, 0.38, 0.38, 0.37, 0.35, 0.36, 0.35, 0.35, 0.34, 0.34, 0.34,
              0.33, 0.32, 0.32, 0.31, 0.3, 0.3, 0.3, 0.29, 0.29, 0.28, 0.28, 0.27, 0.27, 0.26, 0.25]

        self.ampR = rR[sp02.index(spo2value)]
        self.ampIR = IR[sp02.index(spo2value)]

        self.showR.setText(str(round(self.ampR / self.ampIR, 2)))
        self.showRac.setText(str(round(self.ampR, 2)))
        self.showIRac.setText(str(round(self.ampIR, 2)))
        self.showDC.setText(str(round(self.minR, 2)))

        self._initial(self.ampR, self.ampIR)

        self._Rplt.clear()
        self._Rplt.plot(x=list(R), y=list(sp02), pen=self.plot_colors[0])
        self._Rplt.plot(x=list(Ri), y=list(sp02), pen=self.plot_colors[0], symbolBrush=(255, 0, 0))

    def _initial(self, amp_r, amp_ir):
        """
        contain the initial figure in the UI
        """
        hr = self.HRsl.value()
        self.showHR.setText(str(hr))

        t = np.linspace(0.2, 0.8, 100)
        s_r, s_ir = self.ppg_parameters(self.minR, amp_r, self.minIR, amp_ir, t, 200)

        data = sio.loadmat('curvesHB')
        x = data['x']

        xhb_o2, hb_o2, x_oxy_hb, oxy_hb = x[0], x[1], x[2], x[3]

        spo2value = self.spo2sl.value()
        hb_x = np.linspace((700 - spo2value), 1000, 100)
        hb_y = x[5]

        self._plt.clear()
        self._plt.plot(x=list(t)[-PLOT_UPDATE_POINTS:], y=list(s_r)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[1])
        self._plt.plot(x=list(t)[-PLOT_UPDATE_POINTS:], y=list(s_ir)[-PLOT_UPDATE_POINTS:], pen=self.plot_colors[0])

        self._Rplt_2.clear()
        self._Rplt_2.plot(x=list(xhb_o2), y=list(hb_o2), pen=self.plot_colors[0])
        self._Rplt_2.plot(x=list(x_oxy_hb), y=list(oxy_hb), pen=self.plot_colors[0])
        self._Rplt_2.plot(x=list(hb_x), y=list(hb_y), pen=self.plot_colors[1])

    def _update_plot(self):
        """
        Updates and redraws the graphics in the plot.
        """
        # Getting heart rate
        hr = float(self.HRsl.value())
        # generates the time
        self.tPPG = time() - self.timestamp
        self.sR, self.sIR = self.ppg_parameters(self.minR, self.ampR, self.minIR, self.ampIR, self.tPPG, hr)

        # store data into variables 
        self.TIME.append(self.tPPG)
        self.dataR.append(self.sR)
        self.dataIR.append(self.sIR)

        # Draw new data
        self._plt_2.clear()
        self._plt_2.plot(x=list(self.TIME)[-PLOT_UPDATE_POINTS:], y=list(self.dataR)[-PLOT_UPDATE_POINTS:],
                         pen=self.plot_colors[1])
        self._plt_2.plot(x=list(self.TIME)[-PLOT_UPDATE_POINTS:], y=list(self.dataIR)[-PLOT_UPDATE_POINTS:],
                         pen=self.plot_colors[0])

    ## -----------------------------------
    ##      Window Configuration
    ## -----------------------------------
    def _configure_plot(self):
        """
        Configures specific elements of the PyQtGraph plots.
        """
        self.plt.setBackground(background=None)
        self.plt.setAntialiasing(True)
        self._plt = self.plt.addPlot(row=1, col=1)
        self._plt.setLabel('bottom', "Tiempo", "s")
        self._plt.setLabel('left', "Amplitud", "Volt")
        self._plt.showGrid(x=False, y=True)

        self.plt_2.setBackground(background=None)
        self.plt_2.setAntialiasing(True)
        self._plt_2 = self.plt_2.addPlot(row=1, col=1)
        self._plt_2.setLabel('bottom', "Tiempo", "s")
        self._plt_2.setLabel('left', "Amplitud", "Volt")
        self._plt_2.showGrid(x=False, y=False)

        self.Rplt.setBackground(background=None)
        self.Rplt.setAntialiasing(True)
        self._Rplt = self.Rplt.addPlot(row=1, col=1)
        self._Rplt.setLabel('bottom', "R-value")
        self._Rplt.setLabel('left', "%SpO2")
        self._Rplt.showGrid(x=False, y=True)

        self.Rplt_2.setBackground(background=None)
        self.Rplt_2.setAntialiasing(True)
        self._Rplt_2 = self.Rplt_2.addPlot(row=1, col=1)
        self._Rplt_2.setLabel('bottom', "Longitud de Onda", "*nm")
        self._Rplt_2.setLabel('left', "Absorbancia")
        self._Rplt_2.showGrid(x=False, y=True)

    def setupUI(self):
        """
       Configures everything regarding the UI
        """
        # Default Heart Rate configuration
        self.HRsl.setMaximum(250)
        self.HRsl.setMinimum(50)
        self.HRsl.setValue(80)
        self.HRsl.setTickPosition(QSlider.TicksBelow)
        self.HRsl.setTickInterval(1)
        # Default SpO2 configuration
        self.spo2sl.setMaximum(100)
        self.spo2sl.setMinimum(50)
        self.spo2sl.setValue(100)
        self.spo2sl.setTickPosition(QSlider.TicksBelow)
        self.spo2sl.setTickInterval(1)

    def _enable_ui(self, enabled):
        """
        Enable touching the buttons in the UI
        """
        self.startButton.setEnabled(enabled)
        self.spo2sl.setEnabled(enabled)
        self.HRsl.setEnabled(enabled)
        self.showSpo2.setEnabled(enabled)
        self.stopButton.setEnabled(not enabled)

    def buttons(self):
        """
        Configures the connections between signals and UI elements.
        """
        self.startButton.clicked.connect(self.start)
        self.stopButton.clicked.connect(self.stop)

    def start(self):
        """
        This function works when the start button is clicked
        It generates a t0 time and activates the Qt timer which connects to update_plot
        :return:
        """
        self._plt_2.clear()
        self._enable_ui(False)
        self.timestamp = time()
        self._timer_plot.start(PLOT_UPDATE_TIME)

    def stop(self):
        """
        This function works when the stop button is clicked
        it stop the timer and resets the buffers
        """
        self._initial(self.ampR, self.ampIR)
        self._enable_ui(True)
        self._timer_plot.stop()
        self.reset_buffers()

    def _configure_timers(self):
        """
        Configures specific elements of the QTimers.
        :return:
        """
        self._timer_plot = QtCore.QTimer(self)  # gives _timer_plot the attribute of QtCore.QTimer
        self._timer_plot.timeout.connect(self._update_plot)  # connects with _update_plot method

    def reset_buffers(self):
        """
        Clear everything into the vectors that have the signals
        """
        self.dataR.clear()
        self.dataIR.clear()
        self.TIME.clear()
# -----------------------------------------------------------------------------
#                             App Execution
# -----------------------------------------------------------------------------        
# Instance to start an application in windows
app = QApplication(sys.argv)
# Creating an object for the class created above
_mainWindow = mainWindow()
_mainWindow.show()
# Running application
app.exec_()
